File: ios/AnyProvCore/Sources/AnyProvCore/Sync/E2BKeyClient.swift
import Foundation

/// Errors surfaced when the user sets / clears their E2B API key from
/// outside an active coding session (Settings → E2B key). Mirrors the
/// `SkillsMCPError` shape so callers can present a consistent UI.
public enum E2BKeyError: Error, LocalizedError {
    case noServer
    case serverError(String)
    case timeout

    public var errorDescription: String? {
        switch self {
        case .noServer: return "Not paired with a desktop server."
        case let .serverError(msg): return msg
        case .timeout: return "Desktop did not acknowledge the E2B key change."
        }
    }
}

/// Result of an `e2b_set_key` round trip. The desktop responds with
/// `e2b_key_ack`, which carries whether the per-connection override is
/// now active. We surface that so the Settings card can update its label
/// without a follow-up query.
public struct E2BKeyResult: Equatable, Sendable {
    public let overrideActive: Bool
    public init(overrideActive: Bool) {
        self.overrideActive = overrideActive
    }
}

/// Out-of-session client for the `e2b_set_key` round trip. The Sandboxes
/// view also has its own copy (via `SandboxesStore`) so live `e2b_log`
/// frames can be observed; this one is for the Settings-level entry
/// point that should not have to open the full Sandboxes sheet.
///
/// Each call opens a short-lived WebSocket, sends `e2b_set_key`, waits
/// for the matching `e2b_key_ack` (or an `error`), then disconnects. The
/// desktop stores the override in memory against the bearer token only,
/// so re-pairing clears it.
public final class E2BKeyClient: @unchecked Sendable {
    public init() {}

    /// Send a new override (or clear it with an empty string) and return
    /// whether the override is active.
    public func setKey(
        _ apiKey: String,
        endpoint: ServerClient.Endpoint,
        token: String,
        timeoutSeconds: TimeInterval = 10
    ) async throws -> E2BKeyResult {
        let client = ServerClient()
        var opError: Error = E2BKeyError.timeout
        var result = E2BKeyResult(overrideActive: !apiKey.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
        do {
            let stream = try await client.connect(endpoint: endpoint, token: token)
            try await client.send(.e2bSetKey(apiKey: apiKey))
            let deadline = Date().addingTimeInterval(timeoutSeconds)
            for await message in stream {
                switch message {
                case let .e2bKeyAck(active):
                    result = E2BKeyResult(overrideActive: active)
                    await client.disconnect()
                    return result
                case let .error(_, text):
                    opError = E2BKeyError.serverError(
                        text.isEmpty ? "Desktop rejected the E2B key change." : text
                    )
                default:
                    break
                }
                if Date() > deadline { break }
            }
            await client.disconnect()
        } catch {
            opError = error
        }
        throw opError
    }
}

File: android/RoamSocketCore/src/main/kotlin/app/roamsocket/core/sandboxes/E2BKeyClient.kt
/*
 * E2B key wire client. Mirrors the iOS
 * `ios/AnyProvCore/.../Sync/E2BKeyClient.swift` (the one that ships
 * with PR #106).
 *
 * Each call opens a short-lived WebSocket, sends `e2b_set_key`, waits
 * for the matching `e2b_key_ack` (or an `error`), then disconnects.
 * The desktop stores the override in memory against the bearer
 * token only, so re-pairing clears it.
 */
package app.roamsocket.core.sandboxes

import app.roamsocket.core.protocol.ClientMessage
import app.roamsocket.core.protocol.ServerMessage
import app.roamsocket.core.server.Endpoint
import app.roamsocket.core.server.ServerClient
import kotlinx.coroutines.TimeoutCancellationException
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withTimeoutOrNull

/** Errors raised by the E2B key round trip. */
public sealed class E2BKeyError(message: String) : RuntimeException(message) {
    public data object NoServer : E2BKeyError("Not paired with a desktop server.")
    public data class ServerError(public val detail: String) :
        E2BKeyError(detail.ifEmpty { "Desktop rejected the E2B key change." })
    public data object Timeout : E2BKeyError("Desktop did not acknowledge the E2B key change.")
}

/** Result of an `e2b_set_key` round trip — mirrors the desktop's
 *  `e2b_key_ack` payload. */
public data class E2BKeyResult(
    val overrideActive: Boolean,
)

public class E2BKeyClient(
    private val serverClient: ServerClient = ServerClient(),
) {
    public suspend fun setKey(
        apiKey: String,
        endpoint: Endpoint,
        token: String,
        timeoutSeconds: Long = 10,
    ): E2BKeyResult {
        val session = serverClient.connect(endpoint = endpoint, token = token)
        try {
            session.send(ClientMessage.E2bSetKey(apiKey = apiKey))
            val deadline = System.currentTimeMillis() + timeoutSeconds * 1000
            while (true) {
                val remaining = deadline - System.currentTimeMillis()
                if (remaining <= 0) throw E2BKeyError.Timeout
                val next = withTimeoutOrNull(remaining) {
                    session.incoming.first()
                } ?: throw E2BKeyError.Timeout
                when (next) {
                    is ServerMessage.E2bKeyAck -> {
                        return E2BKeyResult(overrideActive = next.overrideActive)
                    }
                    is ServerMessage.Error -> {
                        throw E2BKeyError.ServerError(next.message)
                    }
                    else -> continue
                }
            }
            @Suppress("UNREACHABLE_CODE")
            error("unreachable")
        } catch (e: E2BKeyError) {
            throw e
        } catch (e: TimeoutCancellationException) {
            throw E2BKeyError.Timeout
        } catch (e: Exception) {
            throw E2BKeyError.ServerError(e.message ?: e.javaClass.simpleName)
        } finally {
            runCatching { session.close() }
        }
    }
}
